kedro/framework/cli/starters.py

```diff
@@ -543,64 +542,51 @@ def _get_extra_context(  # noqa: PLR0913
             None in case the flag wasn't used
 
     Returns:
-        the prompts_required dictionary, with all the redundant information removed.
+        Config dictionary, passed the necessary processing, with default values if needed.
     """
     if config_path:
-        extra_context = _fetch_config_from_file(config_path)
-        _validate_config_file_against_prompts(extra_context, prompts_required)
-        _validate_config_file_inputs(extra_context, starter_alias)
+        extra_context = _fetch_validate_parse_config_from_file(
+            config_path, prompts_required, starter_alias
+        )
     else:
-        extra_context = _fetch_config_from_user_prompts(
+        extra_context = _fetch_validate_parse_config_from_user_prompts(
             prompts_required, cookiecutter_context
         )
 
-    # Format
-    extra_context.setdefault("kedro_version", version)
-
-    converted_tools = _convert_tool_names_to_numbers(selected_tools)
-
-    if converted_tools is not None:
-        extra_context["tools"] = converted_tools
-
+    # Update extra_context, if CLI inputs are available
+    if selected_tools is not None:
+        tools_numbers = _convert_tool_short_names_to_numbers(selected_tools)
+        extra_context["tools"] = _convert_tool_numbers_to_readable_names(tools_numbers)
     if project_name is not None:
         extra_context["project_name"] = project_name
+    if example_pipeline is not None:
+        extra_context["example_pipeline"] = str(_parse_yes_no_to_bool(example_pipeline))
 
-    # Map the selected tools lists to readable name
-    tools_context = extra_context.get("tools")
-    tools = _parse_tools_input(tools_context)
-
-    # Check if no tools selected
-    if not tools:
-        extra_context["tools"] = str(["None"])
-    else:
-        extra_context["tools"] = str([NUMBER_TO_TOOLS_NAME[tool] for tool in tools])
-
-    extra_context["example_pipeline"] = _parse_yes_no_to_bool(
-        example_pipeline
-        if example_pipeline is not None
-        else extra_context.get("example_pipeline", "no")
-    )
+    # set defaults for required fields, will be used mostly for starters
+    extra_context.setdefault("kedro_version", version)
+    extra_context.setdefault("tools", str(["None"]))
```

Comment:
`str(["None"])` still seems like a bit of an odd value to me.. We don't have to change that in this PR, but I think it would be good to discuss.

Reply:
Yes, using .get() is indeed a safer method for retrieving values from a dictionary. However, I noticed we were overusing it in every step of our processing. This made the code less precise and harder to comprehend, particularly in discerning whether and where default values were being provided. 
Upon analyzing the code, I realized the default values were actually set in the _get_extra_context() function. During the refactoring of this function, I made the inclusion of default values more explicit by introducing a 'default' section and provided comments about it in DocString. Subsequently, I replaced all instances of variable access using get() with direct access. This change aims to enhance the clarity and strictness of our code. 
Now, I hope it's clear that values must be specified, and to understand which values are used, one should review the previous processing steps instead of worrying about default values at each stage.

Reply:
Idea, just noting this here but not adding to the scope of the PR - could we save it as an unwrapped list `["None"]` instead?

Context:

The reason we pass through a list wrapped in a string is because `cookiecutter` interprets a list as a provision of options, from which it selects one to save. To preserve all the chosen tools, we wrap the list in a string. Though the option for no tools is processed separately, `pyproject.toml` has no knowledge of what is passed through and so we cannot pass through just the string "None" (which will be unwrapped).

However, none tools cannot be selected with any other tools, and so passing through just a list shouldn't have any adverse effects; `cookiecutter` would pick the only option from the list - "None" - and pass that through to `pyproject.toml`. 

Would we run into the same problem with unwrapping strings? If so we could always pass through `[["None"]]` but it doesn't seem much better than `str(["None"])` 🤔 

Additionally, for telemetry, should we be consistent with what we store (a list for several tools, but this would save a string if no tools are selected)?

Alternatively, I believe using dictionaries to pass a collection of objects through `cookiecutter` is possible, but it would require significant changes to pyproject.toml - I'm unsure we'd be able to preserve the list type as we'd have to somehow process these values within the `cookiecutter` execution.

Any opinions on this? CC: @DimedS @merelcht

Reply:
@AhdraMeraliQB , and why we need a string of a list? Why just a string is not enough, like:
tools = "Linting, Testing, ..." or tools = "None"

Reply:
@DimedS 
If I remember correctly, the reason is tied to telemetry - it's easier (and more intuitive) to query tools data when stored as a list than as a string. We did originally pass this through as a string ([xref](https://github.com/kedro-org/kedro/pull/3188/files#:~:text=cookiecutter.kedro_version%20%7D%7D%22-,add_ons,-%3D%20%7B%7B%20cookiecutter.add_ons)), but was suggested to be changed as part of #3264 and in #3333

The fix in #3426 introduced using [str["None"]](https://github.com/kedro-org/kedro/pull/3426/files#:~:text=else%3A-,extra_context,-%5B%22tools%22) to workaround `cookiecutter` limitations ([see Nok's comment](https://github.com/cookiecutter/cookiecutter/issues/1975#issuecomment-1805044493)) and the rest is history 😅